File: backend_fastapi/app/services/push.py
import json
import os
import asyncio
from app.db.mongodb import get_db

VAPID_PUBLIC_KEY = os.getenv("VAPID_PUBLIC_KEY")
VAPID_PRIVATE_KEY = os.getenv("VAPID_PRIVATE_KEY")
VAPID_SUBJECT = os.getenv("VAPID_SUBJECT", "mailto:safety@saferoute.app")

async def broadcast_alert(active_websockets, trip_id: str, alert_payload: dict):
    db = get_db()

    # 1. Broadcast over active WebSockets for open tabs
    room_sockets = active_websockets.get(trip_id, [])
    for ws in list(room_sockets):
        try:
            await ws.send_json({"type": "alert:new", "data": alert_payload})
        except Exception:
            pass

    # 2. Query push subscriptions for closed-tab system notifications
    trip = db["trips"].find_one({"_id": trip_id})
    if not trip:
        return

    user_id = trip["user_id"]
    subscriptions = db["push_subscriptions"].find({"user_id": user_id})

    if VAPID_PUBLIC_KEY and VAPID_PRIVATE_KEY:
        try:
            from pywebpush import webpush, WebPushException
            for sub in subscriptions:
                try:
                    webpush(
                        subscription_info={
                            "endpoint": sub["endpoint"],
                            "keys": {"p256dh": sub["p256dh"], "auth": sub["auth"]}
                        },
                        data=json.dumps(alert_payload),
                        vapid_private_key=VAPID_PRIVATE_KEY,
                        vapid_claims={"sub": VAPID_SUBJECT}
                    )
                except WebPushException as ex:
                    if ex.response and ex.response.status_code in (404, 410):
                        db["push_subscriptions"].delete_one({"_id": sub["_id"]})
        except Exception as e:
            print("WebPush error:", e)

File: backend/src/services/push.js
import webpush from 'web-push';

if (process.env.VAPID_PUBLIC_KEY && process.env.VAPID_PRIVATE_KEY && process.env.VAPID_SUBJECT) {
  webpush.setVapidDetails(
    process.env.VAPID_SUBJECT,
    process.env.VAPID_PUBLIC_KEY,
    process.env.VAPID_PRIVATE_KEY
  );
}

export const sendAlert = async (io, db, tripId, alertPayload) => {
  io.to(`trip:${tripId}`).emit('alert:new', alertPayload);

  const trip = db.prepare('SELECT user_id FROM trips WHERE id = ?').get(tripId);
  if (!trip) return;

  const subscriptions = db.prepare('SELECT * FROM push_subscriptions WHERE user_id = ?').all(trip.user_id);
  
  for (const sub of subscriptions) {
    const pushSub = {
      endpoint: sub.endpoint,
      keys: {
        p256dh: sub.p256dh,
        auth: sub.auth
      }
    };

    try {
      await webpush.sendNotification(pushSub, JSON.stringify(alertPayload));
    } catch (err) {
      if (err.statusCode === 410 || err.statusCode === 404) {
        db.prepare('DELETE FROM push_subscriptions WHERE id = ?').run(sub.id);
      } else {
        console.error('Push notification error:', err);
      }
    }
  }
};